### ISSUE TITLE
Building different targets on same Dockerfile causes a Duplicate resource URN error

### ISSUE THREAD

**Issue by duro** [pulumi/pulumi-awsx#1171]
### What happened?

I have a Dockerfile with two targets in it (migrator and production). I need to build both of these targets on every pulumi up. So I have two instances of awsx.acr.Image which are very similar, but have different names, and use different targets.

However, when I do this, I get the error "Duplicate resource URN"

This seems to be related to how the `createUniqueImageName` function works here: https://github.com/pulumi/pulumi-awsx/blob/52e698db4d85d7fd4fa292027d04766ce4b1ceba/awsx/ecr/image.ts#L93

Since all it considers is context, dockerfile, and args (and not resource name, or target) these two resource definitions end up with the same image name.

### Example

```ts
export const migratorRepo = new awsx.ecr.Repository(`${stackName}-migrator`, {
  lifecyclePolicy: {
    rules: [{ maximumNumberOfImages: 25, tagStatus: 'any' }],
  },
});

export const appsRepo = new awsx.ecr.Repository(`${stackName}-api`, {
  lifecyclePolicy: {
    rules: [{ maximumNumberOfImages: 25, tagStatus: 'any' }],
  },
});

const migratorImage = new awsx.ecr.Image(PREFIX + 'migrator', {
  repositoryUrl: migratorRepo.url,
  context: path.join(__dirname, '../../../../'),
  dockerfile: path.join(__dirname, '../../', 'Dockerfile'),
  platform: 'linux/amd64',
  target: 'migration',
  cacheFrom: ["base"],
  args: {
    BUILDKIT_INLINE_CACHE: "1",
  },
});

const apiImage = new awsx.ecr.Image(PREFIX + 'api', {
  repositoryUrl: appsRepo.url,
  context: path.join(__dirname, '../../../../'),
  dockerfile: path.join(__dirname, '../../', 'Dockerfile'),
  platform: 'linux/amd64',
  target: 'production',
  cacheFrom: ["base"],
  args: {
    BUILDKIT_INLINE_CACHE: "1",
  },
});
```

### Output of `pulumi about`

CLI          
Version      3.76.1
Go Version   go1.20.6
Go Compiler  gc

Plugins
NAME    VERSION
nodejs  unknown

Host     
OS       darwin
Version  13.6
Arch     arm64

### Additional context

My guess is that at the very least the `createUniqueImageName` function should factor in things like target, platform, and maybe even cacheFrom.

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by antdking** [comment]
I helped out with the same issue in slack recently (very possibly the same person as the issue author).

I'm little confused as to why the underlying `docker.Image` resource isn't using the provided resourceName, either verbatim or as a prefix for the calculated name to work around the URN limitations of parent scoped resources.
This would negate the need for keeping on adding more details into the hash.

In terms of migration, setting an alias should work without requiring a flag.

**Comment by loganyott** [comment]
I had the same thought about resource name not being used. 

In my case, I am using a custom `FargateService` component that passes `imageArgs` down to `awsx.ecs.FargateService` and uses `command` + `entrypoint` to change which service is run. While this helped me realize I can keep that code to one image and pass that instead (yay faster builds) but the duplicate URN message from `awsx.ecr.Image` was confusing until I found this issue.

While I did refactor to avoid the issue, my early "quick fix" was throwing the resource name into `args` to change the hash.
